### ISSUE TITLE
cirq.resolve_parameters is slow

### ISSUE THREAD

**Issue by rht** [quantumlib/Cirq#6091]
**Description of the issue**

The usage context is to run hybrid variational circuit fast. For a small circuit of 2 qubits, consisting of [`ZZFeatureMap`](https://qiskit.org/documentation/stubs/qiskit.circuit.library.ZZFeatureMap.html) and [`RealAmplitude`](https://qiskit.org/documentation/stubs/qiskit.circuit.library.RealAmplitudes.html), I find `cirq.resolve_parameters` to be ~15x slower than if I were to reconstruct the circuit from scratch with concrete numbers replacing the parameters instead.

**How to reproduce the issue**

```python
import time

import numpy as np
import cirq
import sympy


def make_cirq_circuit(
    num_qubits,
    circuit_params,
    qubits,
    cirq_input_params,
    cirq_weight_params,
):
    circuit = cirq.Circuit()
    # Qiskit's ZZFeatureMap
    zz_repetitions = 2
    for i in range(zz_repetitions):
        circuit.append([cirq.H(q) for q in qubits])
        circuit.append(
            [cirq.Z(qubits[i]) ** (2.0 * cirq_input_params[i] / np.pi)]
            for i in range(num_qubits)
        )
        circuit.append(cirq.CX(qubits[0], qubits[1]))
        circuit.append(
            cirq.Z(qubits[1])
            ** (
                (2.0 * (np.pi - cirq_input_params[0]) * (np.pi - cirq_input_params[1]))
                / np.pi
            )
        )
    # Qiskit's RealAmplitude
    count = 0
    for i in range(4):
        circuit.append(cirq.CX(qubits[0], qubits[1]))
        circuit.append(cirq.ry(cirq_weight_params[count * 2]).on(qubits[0]))
        circuit.append(cirq.ry(cirq_weight_params[count * 2 + 1]).on(qubits[1]))
        count += 1
    return circuit


num_qubits = 2
qubits = cirq.LineQubit.range(num_qubits)

# Prepare the parameterized circuit
cirq_input_params = [sympy.Symbol("x0"), sympy.Symbol("x1")]
cirq_weight_params = [sympy.Symbol(f"theta{i}") for i in range(8)]
circuit_params = cirq_input_params + cirq_weight_params
circuit = make_cirq_circuit(
    num_qubits,
    circuit_params,
    qubits,
    cirq_input_params,
    cirq_weight_params,
)

# Resolve params
np.random.seed(42)
p = {p: np.random.random() for p in circuit_params}
tic = time.time()
resolved_circuit = cirq.resolve_parameters(circuit, p)
print("Elapsed cirq.resolve_parameters:", round(time.time() - tic, 3))

# What if we instead construct the circuit from scratch.
cirq_input_params = [np.random.random() for i in range(2)]
cirq_weight_params = [np.random.random() for i in range(8)]
circuit_params = cirq_input_params + cirq_weight_params
tic = time.time()
make_cirq_circuit(
    num_qubits,
    circuit_params,
    qubits,
    cirq_input_params,
    cirq_weight_params,
)
print("Elapsed construct from scratch:", round(time.time() - tic, 3))
```
Result:
```
Elapsed cirq.resolve_parameters: 0.045
Elapsed construct from scratch: 0.003
```

**Cirq version**
You can get the cirq version by printing `cirq.__version__`. From the command line:

```
1.1.0
```

**Comment by rht** [comment]
Past issue: #1899 and its resolution #3205.

With Qiskit:
```
from qiskit import QuantumCircuit
from qiskit.circuit.library import RealAmplitudes, ZZFeatureMap
feature_map = ZZFeatureMap(2, reps=2)
ansatz = RealAmplitudes(2)
qc = QuantumCircuit(2)
qc.compose(feature_map, inplace=True)
qc.compose(ansatz, inplace=True)
tic = time.time()
qc.bind_parameters(p)
print("Elapsed Qiskit:", round(time.time() - tic, 3))
# Elapsed Qiskit: 0.007
```

**Comment by rht** [comment]
Tracking down the slowest part of `cirq.resolve_parameters` for this example, >90% of the time is spent on https://github.com/quantumlib/Cirq/blob/a95f009617bcdc613cfbca169fbedbfea1de85b7/cirq-core/cirq/protocols/resolve_parameters.py#L189, and in the `ParamResolver` class, >90% of the time is spent in this line: https://github.com/quantumlib/Cirq/blob/a95f009617bcdc613cfbca169fbedbfea1de85b7/cirq-core/cirq/study/resolver.py#L187.

**Comment by rht** [comment]
Replacing `import sympy` with `import symengine as sympy` throughout the library, I got
```
Elapsed cirq.resolve_parameters: 0.011
```

**Comment by tanujkhattar** [comment]
@NoureldinYosri  to test if its potentially a library loading issue by rerunning the test multiple times. 

We should also look at `symengine` and see if it breaks any of our tests.

**Comment by NoureldinYosri** [comment]
@tanujkhattar @rht as I expected running the code twice gives a much better performance
```$ python perf.py
Elapsed cirq.resolve_parameters: 0.023
Elapsed cirq.resolve_parameters: 0.02
Elapsed construct from scratch: 0.001
Elapsed construct from scratch: 0.001
```

code is the just same but repeating the timing blocks
```
tic = time.time()
resolved_circuit = cirq.resolve_parameters(circuit, p)
print("Elapsed cirq.resolve_parameters:", round(time.time() - tic, 3))
tic = time.time()
resolved_circuit = cirq.resolve_parameters(circuit, p)
print("Elapsed cirq.resolve_parameters:", round(time.time() - tic, 3))

...

tic = time.time()
make_cirq_circuit(
    num_qubits,
    circuit_params,
    qubits,
    cirq_input_params,
    cirq_weight_params,
)
print("Elapsed construct from scratch:", round(time.time() - tic, 3))
tic = time.time()
make_cirq_circuit(
    num_qubits,
    circuit_params,
    qubits,
    cirq_input_params,
    cirq_weight_params,
)
print("Elapsed construct from scratch:", round(time.time() - tic, 3))
```

it's not enough that the code starts with `import sympy` since that line won't import the whole library but only the parts of the library named in the library level `__init__.py`. but if in our logic we do `from sympy.X import Y` then `Y` won't be imported until the interpeter gets to the first file containing that import line. 

python interpetars do eager imports of the parts of the libraries that they see, however some libraries (e.g. scipy) explicitly do lazy imports to mitigiate that https://peps.python.org/pep-0690/ 

so key takeaway the slowness of the first run is due to the size of the part of `sympy` library that we use. however any run after the first should be fast.

**Comment by skushnir123** [comment]
I can take up this issue.

**Comment by skushnir123** [comment]
@rht I merged this PR https://github.com/quantumlib/Cirq/pull/6128 a couple weeks back to improve performance quite a bit. Let me know if things look good on your end!

**Comment by rht** [comment]
Yeah, it does speed up a lot to be usable for experimentation, thanks a lot for it!

**Comment by rht** [comment]
I did try using `symengine` on top of #6128, but it slowed the elapsed time instead.

**Comment by rht** [comment]
Closed via #6128.
